### PR TITLE
Add http client to the Zeebe java client

### LINKED ISSUE #16538: An Apache HttpClient 5 can be created in the Zeebe Java client
### Description

As part of the ongoing support for a Zeebe REST API, the Zeebe Java client needs to be expanded to support REST API communication through an HTTP client.

With #16184 the Zeebe team decided to use Apache HttpClient 5.

### AC
- [x] An HTTP client instance is created when the (java) `ZeebeClient` is created.
- [x] A `TopologyRequestImpl` can use the HTTP client to make a REST API communication with the Zeebe Gateway.

### PR DESCRIPTION
**Pull request by koevskinikola**
## Description

Implement an http client in the Zeebe Java client that will be used to communicate via REST with the REST API server in the Zeebe Gateway.

Notes:
- Client test infrastructure and coverage for REST communication will be added with https://github.com/camunda/zeebe/issues/16643
- QA tests for REST calls will be added with https://github.com/camunda/zeebe/issues/16558

## Related issues

closes #16538 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/ZeebeClientBuilderImpl.java [hunk 1]
```diff
@@ -222,6 +230,10 @@
     if (properties.containsKey(ClientProperties.GATEWAY_ADDRESS)) {
       gatewayAddress(properties.getProperty(ClientProperties.GATEWAY_ADDRESS));
     }
+    if (properties.containsKey(ClientProperties.GATEWAY_REST_API_PORT)) {
+      gatewayRestApiPort(
+          Integer.parseInt(properties.getProperty(ClientProperties.GATEWAY_REST_API_PORT)));
```
[the remark this answers is not in the document]
Comment (koevskinikola) [reply zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/ZeebeClientBuilderImpl.java:235]: :information_source:  We ignore these checks in the `ZeebeClientBuilderImpl`.

File: zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/ZeebeClientBuilderImpl.java [hunk 2]
```diff
@@ -490,6 +514,10 @@
       maxMessageSize(DataSizeUtil.parse(Environment.system().get(MAX_MESSAGE_SIZE)));
     }
 
+    if (Environment.system().isDefined(GATEWAY_REST_API_PORT_VAR)) {
+      gatewayRestApiPort(Integer.parseInt(Environment.system().get(GATEWAY_REST_API_PORT_VAR)));
```
[the remark this answers is not in the document]
Comment (koevskinikola) [reply zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/ZeebeClientBuilderImpl.java:518]: :information_source:  We ignore these checks in the `ZeebeClientBuilderImpl`.

**Review by tmetzke — APPROVED** [review]
👍 Awesome first iteration on this 🚀🥇 
💭 I added some thoughts, one of them needs more immediate action (JSON result handling). All of them can go into a follow-up PR though to make progress here.

File: zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/http/JsonAsyncEntityConsumer.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  @Override
+  protected void streamStart(final ContentType contentType) throws IOException {
+    if (contentType == null) {
+      throw new IOException("Expected to parse a JSON response, but no content type detected");
+    }
+
+    if (ContentType.APPLICATION_JSON.isSameMimeType(contentType)) {
+      isResponse = true;
+    } else if (ContentType.APPLICATION_PROBLEM_JSON.isSameMimeType(contentType)) {
+      isResponse = false;
```
[the remark this answers is not in the document]
Comment (koevskinikola) [reply zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/http/JsonAsyncEntityConsumer.java:74]: I've created a follow-up issue for this: https://github.com/camunda/zeebe/issues/16660

File: zeebe/clients/java/src/main/java/io/camunda/zeebe/client/api/command/ProblemException.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+package io.camunda.zeebe.client.api.command;
+
+import io.camunda.zeebe.gateway.protocol.rest.ProblemDetail;
+
+public class ProblemException extends ClientHttpException {
+  private final ProblemDetail details;
+
+  public ProblemException(final int code, final String reason, final ProblemDetail details) {
+    super(code, reason);
+    this.details = details;
```
[the remark this answers is not in the document]
Comment (koevskinikola) [reply zeebe/clients/java/src/main/java/io/camunda/zeebe/client/api/command/ProblemException.java:25]: Let's adjust this with a follow-up issue.

File: zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/http/JsonAsyncEntityConsumer.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      return null;
+    }
+
+    buffer.asParserOnFirstToken();
+
+    if (isResponse) {
+      return JsonEntity.of(json.readValue(buffer.asParserOnFirstToken(), type));
+    }
+
+    return JsonEntity.of(json.readValue(buffer.asParserOnFirstToken(), ProblemDetail.class));
```
[the remark this answers is not in the document]
Comment (koevskinikola) [reply zeebe/clients/java/src/main/java/io/camunda/zeebe/client/impl/http/JsonAsyncEntityConsumer.java:106]: I'll address this with https://github.com/camunda/zeebe/issues/16660 on the REST API server/controller side.

**Pull request merged**